nupic/research/frameworks/vernon/mixins/noise_robustness_test.py

```diff
@@ -0,0 +1,146 @@
+        assert all(
+            0 < noise <= 1 for noise in noise_levels
+        ), "Noise levels must be between (0, 1]"
+        noise_mean = config.get("noise_mean", 0)
+        noise_std = config.get("noise_std", 1.0)
+        default_evaluate_model_func = config.get("evaluate_model_func", evaluate_model)
+        self.evaluate_model = partial(
+            evaluate_model_with_noise,
+            evaluate_model_func=default_evaluate_model_func,
+            noise_levels=noise_levels,
+            noise_mean=noise_mean,
+            noise_std=noise_std,
+        )
+
+    @classmethod
+    def get_execution_order(cls):
+        eo = super().get_execution_order()
+        eo["setup_experiment"].append("NoiseRobustnessTest: Initialize")
+        return eo
+
+
+def evaluate_model_with_noise(
+    model,
+    loader,
+    device,
+    batches_in_epoch=sys.maxsize,
+    criterion=F.nll_loss,
+    complexity_loss_fn=None,
+    active_classes=None,
+    progress=None,
+    post_batch_callback=None,
+    transform_to_device_fn=None,
+    evaluate_model_func=None,
+    noise_levels=None,
+    noise_mean=0,
+    noise_std=1,
+):
+    noise_results = {}
+    dataset_transform = loader.dataset.transform
+    zero_noise_results = evaluate_model_func(
+        model,
+        loader,
+        device,
+        batches_in_epoch=batches_in_epoch,
+        criterion=criterion,
+        complexity_loss_fn=complexity_loss_fn,
+        active_classes=active_classes,
+        progress=progress,
+        post_batch_callback=post_batch_callback,
+        transform_to_device_fn=transform_to_device_fn,
+    )
+    for noise_level in noise_levels:
+        loader.dataset.transform = transforms.Compose(
+            [dataset_transform, GaussianNoise(noise_level, noise_mean, noise_std)]
+        )
+        noise_results[noise_level] = evaluate_model_func(
+            model,
+            loader,
+            device,
+            batches_in_epoch=batches_in_epoch,
+            criterion=criterion,
+            complexity_loss_fn=complexity_loss_fn,
+            active_classes=active_classes,
+            progress=progress,
+            post_batch_callback=post_batch_callback,
+            transform_to_device_fn=transform_to_device_fn,
+        )
+    loader.dataset.transform = dataset_transform
+    all_results = {
+        key + "_" + str(noise_level) + "_noise": results[key]
+        for noise_level, results in noise_results.items()
+        for key in results
+    }
+    all_results.update(zero_noise_results)
+    return all_results
+
+
+def noisy_getitem(idx, get_item_func=None, noise_level=0, noise_mean=0, noise_std=1):
+    data, target = get_item_func(idx)
+    noise_indices = torch.bernoulli(torch.ones_like(data) * noise_level)
+    noise = torch.normal(noise_mean, noise_std, data.shape) * noise_indices
+    return data + noise, target
+
+
+class GaussianNoise:
+    def __init__(self, noise_level, mean, std):
+        self.mean = mean
+        self.std = std
```

Comment:
OCD comment: if you're following [this guideline](https://discuss.pytorch.org/t/how-to-add-noise-to-mnist-dataset-when-using-pytorch/59745) for creating a transform for adding noise, the class should be named according to what it does (e.g., `AddGaussianNoise`). This is also consistent with how transform objects in pytorch are named.

Reply:
Good comment, thanks for pointing this out. Fixed in latest revision